community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/Latch.java

```diff
@@ -0,0 +1,164 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.io.pagecache.impl.muninn;
+
+import java.util.concurrent.locks.LockSupport;
+
+import static org.neo4j.io.pagecache.impl.muninn.UnsafeUtil.getAndSetObject;
+
+/**
+ * This class is similar in many ways to a CountDownLatch(1).
+ *
+ * The main difference is that instances of this specialized Latch implementation are much quicker to allocate and
+ * construct. Each instance also takes up less memory on the heap, and enqueueing wait nodes on the latch is faster.
+ *
+ * There are two reasons why this class is faster to construct: 1. it performs no volatile write during its
+ * construction, and 2. it does not need to allocate an internal Sync object, like CountDownLatch does.
+ */
+final class Latch
+{
+    private static class Node
+    {
+        volatile Node next;
+    }
+
+    private static final class Waiter extends Node
+    {
+        final Thread waitingThread = Thread.currentThread();
+    }
+
+    private static final long stackOffset =
+            UnsafeUtil.getFieldOffset( Latch.class, "stack" );
+    private static final Node end = new Node();
+    private static final Node released = new Node();
+
+    @SuppressWarnings( "unused" )
+    private volatile Node stack; // written to via unsafe
+
+    /**
+     * Release the latch, thereby unblocking all current and future calls to {@link #await()}.
+     */
+    public void release()
+    {
+        // Once the release sentinel is on the stack, it can never (observably) leave.
+        // Waiters might accidentally remove the released sentinel from the stack for brief periods of time, but then
+        // they are required to fix the situation and put it back.
+        // Atomically swapping the release sentinel onto the stack will give us back all the waiters, if any.
+        Node waiters = (Node) getAndSetObject( this, stackOffset, released );
+        if ( waiters == null )
+        {
+            // There are no waiters to unpark, so don't bother.
+            return;
+        }
+        unparkAll( waiters );
+    }
+
+    /**
+     * Wait for the latch to be released, blocking the current thread if necessary.
+     *
+     * This method returns immediately if the latch has already been released.
+     */
+    public void await()
+    {
+        // Put in a local variable to avoid volatile reads we don't need.
+        Node state = stack;
+        if ( state != released )
+        {
+            // The latch hasn't obviously already been released, so we want to add a waiter to the stack. Trouble is,
+            // we might race with release here, so we need to re-check for release after we've modified the stack.
+            Waiter waiter = new Waiter();
+            state = (Node) getAndSetObject( this, stackOffset, waiter );
```

Comment:
That's not a problem, because the reference we null-check on is `static final` so the JIT will constant-fold it and then do dead-code-elimination. Calling through `invokeExact` on `static final` MethodHandles also allows the JIT to inline the target methods as if they were static callsites. In other words, there is no overhead for the current way of doing it.